### PR TITLE
more changes....

### PR DESCRIPTION
**Pull request by muyaooo**
I still cannot run your script on my computer because of the version... so some of the changes might be wrong...

How do you deal with the zero-count trigrams whose pregram also has zero-counts in laplace smoothing? I try to adjust the zero counts to one for trigrams whose pregram has appeared in the training set, but for some trigrams in the test set, their pregrams never appeared in the training set, so such trigrams are not included in the language model... How do you deal with such situation?

### CODE REVIEW

**Comment by Satisian** [comment]
>but for some trigrams in the test set, their pregrams never appeared in the training set, so such trigrams are not included in the language model... How do you deal with such situation?

I'm confused. The current script generates all possible trigrams, including ones not seen on the data set. (At least I think it does!). Without any smoothing they are given a probability of zero. I originally did it so that it only generated the model with observed trigrams (this is what the count_combinatoric function does) but I abandoned that because of the reasons you mentioned.

**Comment by Satisian** [comment]
>I still cannot run your script on my computer because of the version... so some of the changes might be wrong...

As for the version, it's annoying. I really need to convert it to python 2.7 at some point so I can submit it on the DICE machines, but for the moment would you be able to download python 3 to be able to run it? It shouldn't take too long...

**Comment by muyaooo** [comment]
I actually have downloaded the version 3.5 but there are some weird error messages appeared when I am running the program...I will try to fix that later. By the way, the generate_sequence function in my code can run under python 2.7. I borrowed the method from lab2.., which is basically generating bins according to the distribution and a random number between 0 to 1 and see which bins it falls into. I have updated my code so the generate_sequence function should run smoothly now (it is not functioning before). 


As for your code, so you are generating all the possible N-grams and their pregrams and assign the counts of the N-grams and and pregrams that have never appeared in the training set to zero, right? if so, in add-alpha smoothing, the probability for those zero-count N-grams whose pregrams also have zero-count should be 1/(vocabulary size)(correct me if I am wrong). I assign 1/(vocab-size) to all such N-grams and I think my add-alpha model should be the same as yours. But it produces a different perplexity when analysing the training text :( I will check my code again...


I 


I think that works. Howev












At 2016-10-11 08:00:31, "Satisian" <notifications@github.com> wrote:


I still cannot run your script on my computer because of the version... so some of the changes might be wrong...

As for the version, it's annoying. I really need to convert it to python 2.7 at some point so I can submit it on the DICE machines, but for the moment would you be able to download python 3 to be able to run it? It shouldn't take too long...

**Comment by Satisian** [comment]
>I actually have downloaded the version 3.5 but there are some weird error messages appeared when I am running the program...I will try to fix that later. By the way, the generate_sequence function in my code can run under python 2.7. I borrowed the method from lab2.., which is basically generating bins according to the distribution and a random number between 0 to 1 and see which bins it falls into. I have updated my code so the generate_sequence function should run smoothly now (it is not functioning before).

Great. 

>the probability for those zero-count N-grams whose pregrams also have zero-count should be 1/(vocabulary size)(correct me if I am wrong)

Yes, when alpha = 1.

>But it produces a different perplexity when analysing the training text :( I will check my code again...

Hmm... weird. I'll check my code again - there could be some bugs in there. It could also be because the perplexities we calculated were done before we added the #s to the beginning and end of the sentence, so that could have affected it.

**Comment by muyaooo** [comment]
The 1/vocabulary size should apply to every alpha.


The perplexity I calculated for the training.en is 5.98, using the add-one model.




Sent from Mail Master
On 10/11/2016 19:06, Satisian wrote:

I actually have downloaded the version 3.5 but there are some weird error messages appeared when I am running the program...I will try to fix that later. By the way, the generate_sequence function in my code can run under python 2.7. I borrowed the method from lab2.., which is basically generating bins according to the distribution and a random number between 0 to 1 and see which bins it falls into. I have updated my code so the generate_sequence function should run smoothly now (it is not functioning before).

Great.

the probability for those zero-count N-grams whose pregrams also have zero-count should be 1/(vocabulary size)(correct me if I am wrong)

Yes, when alpha = 1.

But it produces a different perplexity when analysing the training text :( I will check my code again...

Hmm... weird. I'll check my code again - there could be some bugs in there. It could also be because the perplexities we calculated were done before we added the #s to the beginning and end of the sentence, so that could have affected it.

**Comment by Satisian** [comment]
>The 1/vocabulary size should apply to every alpha.

Good point. The alphas cancel. I'll check my code. It's probably doing something stupid.

**Comment by Satisian** [comment]
Also, did you ever get around to implementing Good-Turing or any other smoothing methods? If so, could you upload the code.

**Comment by muyaooo** [comment]
I did, and the code is in the script I uploaded. but it gives me so many zeros for large-number occurrence. If we are going to smooth all the Nc+1 that are zero, we have to run a regression of log(N) on c, which will introduce new problems of how to run the right regression... So I decide to drop the GT smoothing. 


I think we can deal with problems like how to decide the best value for alpha. It is more feasible to do than the GT smoothing. What do you think?


Sent from Mail Master
On 10/11/2016 19:50, Satisian wrote:

Also, did you ever get around to implementing Good-Turing or any other smoothing methods? If so, could you upload the code.

**Comment by Satisian** [comment]
>I think we can deal with problems like how to decide the best value for alpha. It is more feasible to do than the GT smoothing. What do you think?

That sounds like a good idea.

Also, I checked my conditional probabilities and they were bugged, so good catch. I've fixed that now, hopefully. The perplexity I calculated for the english language trigram model on the test set was 5.06821887339355. The perplexity I calculated for q1 was 2.3569552537168756. Are they what you got?

Finally, what was our decision on Q3 about which smoothing method was used, and our reasoning? I'm writing up a draft of the report at the moment, so it would be nice to agree on this. Also, I remember you had the maths already written for q1, would you mind sending that over?

**Comment by muyaooo** [comment]
I trained the model on the training.en
the perplexity I calculated for the test set is 8.86, for Q1 is 3.137. Have you updated your code on gibhub? I could have a look at it and see how you calculate it.


For Q3, I look at the model again and I think it could both be laplace smoothing and GT smoothing, because there are other identical items other than the 3.33-e-02. Both the laplace and GT smoothing assign the same possibility for all the zero-occurrences which have the same pregram. I cannot further narrow the range down for now. I think we need more discussion on this question.





--

牛牧遥
北京外国语大学英语学院





At 2016-10-13 02:54:47, "Satisian" <notifications@github.com> wrote:


I think we can deal with problems like how to decide the best value for alpha. It is more feasible to do than the GT smoothing. What do you think?

That sounds like a good idea.

Also, I checked my conditional probabilities and they were bugged, so good catch. I've fixed that now, hopefully. The perplexity I calculated for the english language trigram model on the test set was 5.06821887339355. The perplexity I calculated for q1 was 2.3569552537168756. Are they what you got?

Finally, what was our decision on Q3 about which smoothing method was used, and our reasoning? I'm writing up a draft of the report at the moment, so it would be nice to agree on this. Also, I remember you had the maths already written for q1, would you mind sending that over?

**Comment by muyaooo** [comment]
oh, Beren, it just occurs to me that the model might be calculated using the add-alpha smoothing, because 3.33e-02 equals to 1/30, which is the vocabulary size. The number appears after the trigrams whose pregram may never appear in English(for example, two white spaces), but the gt smoothing will not assign such trigrams this number. So the smoothing method should be add-alpha





--

牛牧遥
北京外国语大学英语学院





At 2016-10-13 02:54:47, "Satisian" <notifications@github.com> wrote:


I think we can deal with problems like how to decide the best value for alpha. It is more feasible to do than the GT smoothing. What do you think?

That sounds like a good idea.

Also, I checked my conditional probabilities and they were bugged, so good catch. I've fixed that now, hopefully. The perplexity I calculated for the english language trigram model on the test set was 5.06821887339355. The perplexity I calculated for q1 was 2.3569552537168756. Are they what you got?

Finally, what was our decision on Q3 about which smoothing method was used, and our reasoning? I'm writing up a draft of the report at the moment, so it would be nice to agree on this. Also, I remember you had the maths already written for q1, would you mind sending that over?

**Comment by muyaooo** [comment]
This is the calculation for Q1...sorry I forgot to attach it in previous emails...

 






--

牛牧遥
北京外国语大学英语学院





At 2016-10-13 02:54:47, "Satisian" <notifications@github.com> wrote:


I think we can deal with problems like how to decide the best value for alpha. It is more feasible to do than the GT smoothing. What do you think?

That sounds like a good idea.

Also, I checked my conditional probabilities and they were bugged, so good catch. I've fixed that now, hopefully. The perplexity I calculated for the english language trigram model on the test set was 5.06821887339355. The perplexity I calculated for q1 was 2.3569552537168756. Are they what you got?

Finally, what was our decision on Q3 about which smoothing method was used, and our reasoning? I'm writing up a draft of the report at the moment, so it would be nice to agree on this. Also, I remember you had the maths already written for q1, would you mind sending that over?

**Comment by Satisian** [comment]
> Have you updated your code on gibhub?

I've uploaded it now.

I think I figured out the problem. I wasn't subtracting 2 from my N for the perplexity, so it was counting the 2#s at the beginning of each line. Now I get 3.13 for Q1 too.

>oh, Beren, it just occurs to me that the model might be calculated using the add-alpha smoothing, because 3.33e-02 equals to 1/30, which is the vocabulary size. The number appears after the trigrams whose pregram may never appear in English(for example, two white spaces), but the gt smoothing will not assign such trigrams this number. So the smoothing method should be add-alpha

yes, I was thinking this too. That's why I wandered to check it with you before going with Good-Turing, which was what we decided before.

**Comment by muyaooo** [comment]
Great. So what is your perplexity for the test set this time? 


Also, I think that for the zero-pregram cases, the smoothing methods we learnt(except interpolation and back off) have no better way to smooth them but assigning 1/vocabulary size....so the answer is still uncertain...


Sent from Mail Master
On 10/12/2016 20:41, Satisian wrote:

Have you updated your code on gibhub?

I've uploaded it now.

I think I figured out the problem. I wasn't subtracting 2 from my N for the perplexity, so it was counting the 2#s at the beginning of each line. Now I get 3.13 for Q1 too.

oh, Beren, it just occurs to me that the model might be calculated using the add-alpha smoothing, because 3.33e-02 equals to 1/30, which is the vocabulary size. The number appears after the trigrams whose pregram may never appear in English(for example, two white spaces), but the gt smoothing will not assign such trigrams this number. So the smoothing method should be add-alpha

yes, I was thinking this too. That's why I wandered to check it with you before going with Good-Turing, which was what we decided before.
